src/test/rpc/Book_test.cpp

```diff
@@ -1072,6 +1072,178 @@ class Book_test : public beast::unit_test::suite
+        //  - Alice and Bob place identical offers for USD -> XRP
+        //  - Charlie places a crossing order that takes both Alice and Bob's
+
+        auto const gw = Account("gateway");
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+        auto const charlie = Account("charlie");
+        auto const USD = gw["USD"];
+
+        env.fund (XRP(1000000), gw, alice, bob, charlie);
+        env.close();
+
+        env (trust(alice, USD(500)));
+        env (trust(bob, USD(500)));
+        env.close();
+
+        env (pay(gw, alice, USD(500)));
+        env (pay(gw, bob, USD(500)));
+        env.close();
+
+        // Alice and Bob offer $500 for 500 XRP
+        env (offer (alice, XRP(500), USD(500)));
+        env (offer (bob, XRP(500), USD(500)));
+        env.close();
+
+        auto wsc = makeWSClient(env.app().config());
+        Json::Value books;
+        {
+            // RPC subscribe to books stream
+            books[jss::books] = Json::arrayValue;
+            {
+                auto& j = books[jss::books].append(Json::objectValue);
+                j[jss::snapshot] = false;
+                j[jss::taker_gets][jss::currency] = "XRP";
+                j[jss::taker_pays][jss::currency] = "USD";
+                j[jss::taker_pays][jss::issuer] = gw.human();
+            }
+
+            auto jv = wsc->invoke("subscribe", books);
+            if (!BEAST_EXPECT(jv[jss::status] == "success"))
+                return;
+        }
+
+        // Charlie places an offer that cross Alice and Charlie's offers
+        env(offer(charlie, USD(1000), XRP(1000)));
+        env.close();
+        env.require(offers(alice, 0), offers(bob, 0), offers(charlie, 0));
+        BEAST_EXPECT(offerOnlyOnceInStream(wsc, 1s, XRP(1000), USD(1000)));
+
+        // RPC unsubscribe
+        auto jv = wsc->invoke("unsubscribe", books);
+        BEAST_EXPECT(jv[jss::status] == "success");
+    }
+
+    void
+    testCrossingMultiBookOffer()
+    {
+        testcase("Crossing multi-book offer");
+
+        // This was added to check that an OfferCreate transaction is only
+        // published once in a stream, even if it auto-bridges across several
+        // books that are under subscription
+
+        using namespace jtx;
+        Env env(*this);
+
+        // Scenario is:
+        //  - Alice has 1 USD and wants 100 XRP
+        //  - Bob has 100 XRP and wants 1 EUR
+        //  - Charlie has 1 EUR and wants 1 USD and should auto-bridge through
+        //    Alice and Bob
+
+        auto const gw = Account("gateway");
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+        auto const charlie = Account("charlie");
+        auto const USD = gw["USD"];
+        auto const EUR = gw["EUR"];
+
+        env.fund(XRP(1000000), gw, alice, bob, charlie);
+        env.close();
+
+        for (auto& account : {alice, bob, charlie})
+        {
+            for (auto& iou : {USD, EUR})
```

Comment:
`account` and `iou` could both be `const&` (tho in code like this I tend not to care - no one is going to be confused if these variables are const or not. Your call if you want to change).